### PR TITLE
rgw: dynamic resharding modifies existing bucket instance

### PR DESCRIPTION
**Pull request by cbodley**
builds on https://github.com/ceph/ceph/pull/35175

* refactors the reshard logic into separate functions to better compose the fault injection and error handling
* adds cleanup of old-style reshards that were in progress
* refactors the test coverage into a test_bucket_reshard() function that's parameterized on the fault to inject

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2020-12-18_18:48:51-rgw-wip-rgw-reshard-instance-distro-basic-smithi/

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
rebased. hoping the unrelated make check failures go away

**Comment by cbodley** [comment]
oh wow, that's an obscure failure. i guess something in 'make check' was linting test_rgw_reshard.py?
```
flake8 run-test: commands[0] | flake8 --select=F,E9 --exclude=venv,.tox
./workunits/rgw/test_rgw_reshard.py:9:1: F401 'pprint.pprint' imported but unused
./workunits/rgw/test_rgw_reshard.py:10:1: F401 're' imported but unused
```
pushed a fix

**Comment by cbodley** [comment]
> 215/215 Test #110: unittest_fault_injector ...................***Timeout 3600.01 sec

pushed a fix for the death tests

**Comment by cbodley** [comment]
jenkins test api

File: src/rgw/rgw_reshard.cc [hunk 1]
```diff
@@ -267,79 +268,183 @@ int RGWBucketReshard::set_resharding_status(rgw::sal::RGWRadosStore* store,
...
+  if (ret = fault.check("revert_target_layout");
+      ret == 0) { // no fault injected, revert the bucket instance metadata
+    ret = store->getRados()->put_bucket_instance_info(bucket_info, false, real_time(), nullptr);
+  }
+
+  if (ret < 0) {
+    ldout(store->ctx(), 0) << "ERROR: " << __func__ << " failed to clear "
+        "target index layout in bucket info: " << cpp_strerror(ret) << dendl;
+
+    bucket_info.layout = std::move(prev); // restore in-memory layout
```
Comment (smanjara) [inline src/rgw/rgw_reshard.cc:383]: after restoring the in memory layout, I think we have to commit it.

**Review by smanjara — CHANGES_REQUESTED** [review]

**Comment by smanjara** [comment]
apart from the one comment, everything else looks great!

**Comment by smanjara** [comment]
everything else looks great!

File: src/rgw/rgw_reshard.cc (hunk 1 above)
Comment (cbodley) [reply src/rgw/rgw_reshard.cc:383]: this block is handling the error from `put_bucket_instance_info()` above, where we failed to commit the updated `bucket_info` to rados. so it's just reverting to what we had before

**Comment by cbodley** [comment]
this merged into https://github.com/ceph/ceph/pull/39002

**Comment by mattbenjamin** [comment]
yay!

**Pull request closed without merging**